### DISCUSSION TITLE
Extend fixed size buffers to all unmanaged types

### DISCUSSION THREAD

**Discussion by rick-de-water** [dotnet/csharplang#5313] [Language Ideas]
[Fixed size buffers](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#fixed-size-buffers) allow you to create a structure where the array is embedded in the struct instead of allocated on the heap. One limitation however is the types that are allowed. Only primitives are allowed (for good reasons), but I think this can be expanded a bit to allow it to work with custom structs as well.

I propose that all [`unmanaged`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/unmanaged-types) are allowed to be used in with `fixed` buffers. All `unmanaged` types are value types with predictable sizes, just like the types that are currently allowed.

This would then also make it possible to use fixed arrays in generic code:
```csharp
public unsafe struct Vector3<T>
    where T : unmanaged
{
    public fixed T values[3];
}
```

I think this would be a great addition for making structs with a very specific memory layout.

**Comment by ghost** [comment]
https://github.com/dotnet/csharplang/issues/1314

**Reply by rick-palmsens** [reply]
Sad to hear that its been on the waiting list for more than 3 years.
